mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/OSKView.m

```diff
@@ -613,114 +622,136 @@ - (void)resetKeyLabels {
   }
 }
 
-- (void)setShiftState:(BOOL)shiftState {
-  if (_shiftState != shiftState) {
-    _shiftState = shiftState;
-    KeyView *shiftKeyL = (KeyView *)[self viewWithTag:MVK_LEFT_SHIFT|0x1000];
-    KeyView *shiftKeyR = (KeyView *)[self viewWithTag:MVK_RIGHT_SHIFT|0x1000];
-    if (shiftState) {
-      [self setKeyLabels:YES alt:self.oskAltState ctrl:self.oskCtrlState];
-      [shiftKeyL setKeyPressed:YES];
-      [shiftKeyR setKeyPressed:YES];
+- (void)setPhysicalShiftState:(BOOL)state {
+  if (_physicalShiftState != state) {
+    _physicalShiftState = state;
+
+    /**
+     * When setting the physical shift state, clear the OSK shift state
+     * The state of the physical key overrides the modifier key clicked in the OSK.
+     */
+    _oskShiftState = NO;
+    
+    if (state) {
+      os_log_debug([KMELogs oskLog], "hardware shift key pressed");
+      [self displayKeyLabelsForLayer];
+      [self displaysShiftKeysAsPressed:YES];
     }
```

Comment:
` [self displayKeyLabelsForLayer];` occurs in the if and the else block, Can this be moved outside the if/else? 
The same happens in functions: setOskShiftState, setPhysicalOptionState, setOskOptionState, setPhysicalControlState, setOskControlState

Reply:
I was considering doing that for the three physical state changes -- the osk state changes are not a simple if/else -- but I am holding off for now. I am about to work on another PR where I may be refactoring this further. I'd like to move this logic elsewhere as it shouldn't be located in the view class.